### PR TITLE
Graal fix multithreading problem

### PR DESCRIPTION
**Pull request by qparis**

### CODE REVIEW

**Comment by plata** [comment]
I'm wondering if this works also without Nashorn compatibility or if we should better use array instead.

**Comment by qparis** [comment]
?

**Comment by qparis** [comment]
Not sure to understand correctly

**Comment by plata** [comment]
Check the linked pr.

**Comment by qparis** [comment]
I think an ArrayList is better. The thing is that you are implicitely doing exactly the same thing by reconverting a List<String> to a String[]. 

When you do this, the Graal implementation of List (PolyglotList) cannot be kept because an Array does not store any method. 

At the end, your solution works because the JS object is copied by the engine, which is basically what I am doing.

**Comment by plata** [comment]
You should add a `fixes #1992`. I'm still not 100% sure that the List <-> Array conversion works like we expect from Nashorn everywhere.

**Comment by Zemogiter** [comment]
Now when I click on the "change engine version" button the installation shows up but the selection does not. I supouse this should look like version list in DXVK.
Zrzut ekranu z 2019-06-11 20-51-35

**Comment by qparis** [comment]
Where are we using Nashorn?

**Comment by Zemogiter** [comment]
Also found this in terminal:
```
Exception in thread "JavaFX Application Thread" java.lang.IllegalStateException: Multi threaded access requested by thread Thread[JavaFX Application Thread,5,main] but is not allowed for language(s) js.
	at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotContextImpl.throwDeniedThreadAccess(PolyglotContextImpl.java:644)
	at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotContextImpl.checkAllThreadAccesses(PolyglotContextImpl.java:562)
	at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotContextImpl.enterThreadChanged(PolyglotContextImpl.java:481)
	at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotContextImpl.enter(PolyglotContextImpl.java:442)
	at org.graalvm.truffle/com.oracle.truffle.polyglot.HostToGuestRootNode.execute(HostToGuestRootNode.java:82)
	at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.callProxy(OptimizedCallTarget.java:325)
	at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.callRoot(OptimizedCallTarget.java:315)
	at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.callBoundary(OptimizedCallTarget.java:302)
	at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.doInvoke(OptimizedCallTarget.java:284)
	at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.callIndirect(OptimizedCallTarget.java:227)
	at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.call(OptimizedCallTarget.java:214)
	at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotList.size(PolyglotList.java:123)
	at java.base/java.util.AbstractList$Itr.hasNext(AbstractList.java:364)
	at java.base/java.util.AbstractCollection.addAll(AbstractCollection.java:351)
	at javafx.base/javafx.collections.ModifiableObservableListBase.addAll(ModifiableObservableListBase.java:99)
	at javafx.base/javafx.collections.FXCollections.observableArrayList(FXCollections.java:356)
	at org.phoenicis.javafx.views.scriptui.StepRepresentationMenu.drawStepContent(StepRepresentationMenu.java:50)
	at org.phoenicis.javafx.views.scriptui.AbstractStepRepresentationWithHeader.installStep(AbstractStepRepresentationWithHeader.java:77)
	at org.phoenicis.javafx.views.scriptui.StepRepresentationMessage.installStep(StepRepresentationMessage.java:28)
	at org.phoenicis.javafx.views.scriptui.SetupUiJavaFXImplementation.showMenuStep(SetupUiJavaFXImplementation.java:117)
	at org.phoenicis.scripts.wizard.UiSetupWizardImplementation.lambda$menu$8(UiSetupWizardImplementation.java:259)
	at org.phoenicis.scripts.ui.UiMessageSender.lambda$runAndWait$0(UiMessageSender.java:30)
	at org.phoenicis.javafx.UiMessageSenderJavaFXImplementation.lambda$run$0(UiMessageSenderJavaFXImplementation.java:42)
	at javafx.graphics/com.sun.javafx.application.PlatformImpl.lambda$runLater$10(PlatformImpl.java:428)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at javafx.graphics/com.sun.javafx.application.PlatformImpl.lambda$runLater$11(PlatformImpl.java:427)
	at javafx.graphics/com.sun.glass.ui.InvokeLaterDispatcher$Future.run(InvokeLaterDispatcher.java:96)
	at javafx.graphics/com.sun.glass.ui.gtk.GtkApplication._runLoop(Native Method)
	at javafx.graphics/com.sun.glass.ui.gtk.GtkApplication.lambda$runLoop$11(GtkApplication.java:277)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

**Comment by qparis** [comment]
This should be fixed with the merge request #1994

File: phoenicis-scripts/src/main/java/org/phoenicis/scripts/wizard/UiSetupWizardImplementation.java [hunk 1]
```diff
@@ -256,7 +257,8 @@ public MenuItem menu(String textToShow, List<String> menuItems) {
      */
     @Override
     public MenuItem menu(String textToShow, List<String> menuItems, String defaultValue) {
-        return messageSender.runAndWait(message -> setupUi.showMenuStep(message, textToShow, menuItems, defaultValue));
+        final List<String> copiedMenuItems = new ArrayList<>(menuItems);
```
Comment (madoar) [inline phoenicis-scripts/src/main/java/org/phoenicis/scripts/wizard/UiSetupWizardImplementation.java]: I think we should replace our `new ArrayList<>(...)` constructs with `List.copyOf(...)` especially when the list is immutable.

Comment (qparis) [reply phoenicis-scripts/src/main/java/org/phoenicis/scripts/wizard/UiSetupWizardImplementation.java]: Done

Comment (Zemogiter) [reply phoenicis-scripts/src/main/java/org/phoenicis/scripts/wizard/UiSetupWizardImplementation.java]: That did not change anything.

Comment (madoar) [reply phoenicis-scripts/src/main/java/org/phoenicis/scripts/wizard/UiSetupWizardImplementation.java]: Then the `import java.util.ArrayList;` is not required anymore.

Comment (qparis) [reply phoenicis-scripts/src/main/java/org/phoenicis/scripts/wizard/UiSetupWizardImplementation.java]: @Zemogiter I have retried and it works perfectly. Are you sure that you are running the right code?

Comment (Zemogiter) [reply phoenicis-scripts/src/main/java/org/phoenicis/scripts/wizard/UiSetupWizardImplementation.java]: I'm positive. I've copied the changes from the latest commit in this PR to the local source and compiled it then created a deb file and installed it.

Comment (qparis) [reply phoenicis-scripts/src/main/java/org/phoenicis/scripts/wizard/UiSetupWizardImplementation.java]: with mvn clean install?

Comment (qparis) [reply phoenicis-scripts/src/main/java/org/phoenicis/scripts/wizard/UiSetupWizardImplementation.java]: Can you try mvn exec:java inside phoencis-javafx directory?

Comment (Zemogiter) [reply phoenicis-scripts/src/main/java/org/phoenicis/scripts/wizard/UiSetupWizardImplementation.java]: I was using mvn clean install as usual. And when used mvn exec:java I got this:
```
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.google.inject.internal.cglib.core.$ReflectUtils$1 (file:/usr/share/maven/lib/guice.jar) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain)
WARNING: Please consider reporting this to the maintainers of com.google.inject.internal.cglib.core.$ReflectUtils$1
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
[INFO] Scanning for projects...
Downloading from central: https://repo.maven.apache.org/maven2/org/codehaus/mojo/versions-maven-plugin/2.7/versions-maven-plugin-2.7.pom
Downloaded from central: https://repo.maven.apache.org/maven2/org/codehaus/mojo/versions-maven-plugin/2.7/versions-maven-plugin-2.7.pom (17 kB at 23 kB/s)
Downloading from central: https://repo.maven.apache.org/maven2/org/codehaus/mojo/versions-maven-plugin/2.7/versions-maven-plugin-2.7.jar
Downloaded from central: https://repo.maven.apache.org/maven2/org/codehaus/mojo/versions-maven-plugin/2.7/versions-maven-plugin-2.7.jar (310 kB at 307 kB/s)
Downloading from central: https://repo.maven.apache.org/maven2/org/codehaus/mojo/exec-maven-plugin/1.6.0/exec-maven-plugin-1.6.0.pom
Downloaded from central: https://repo.maven.apache.org/maven2/org/codehaus/mojo/exec-maven-plugin/1.6.0/exec-maven-plugin-1.6.0.pom (13 kB at 81 kB/s)
Downloading from central: https://repo.maven.apache.org/maven2/org/codehaus/mojo/exec-maven-plugin/1.6.0/exec-maven-plugin-1.6.0.jar
Downloaded from central: https://repo.maven.apache.org/maven2/org/codehaus/mojo/exec-maven-plugin/1.6.0/exec-maven-plugin-1.6.0.jar (58 kB at 326 kB/s)
[INFO] 
[INFO] -------------------< org.phoenicis:phoenicis-javafx >-------------------
[INFO] Building phoenicis-javafx 5.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
Downloading from jitpack.io: https://jitpack.io/org/phoenicis/phoenicis-configuration/5.0-SNAPSHOT/maven-metadata.xml
Downloading from jitpack.io: https://jitpack.io/org/phoenicis/phoenicis-configuration/5.0-SNAPSHOT/phoenicis-configuration-5.0-SNAPSHOT.pom
[WARNING] The POM for org.phoenicis:phoenicis-configuration:jar:5.0-SNAPSHOT is missing, no dependency information available
Downloading from jitpack.io: https://jitpack.io/org/phoenicis/phoenicis-repository/5.0-SNAPSHOT/maven-metadata.xml
Downloading from jitpack.io: https://jitpack.io/org/phoenicis/phoenicis-repository/5.0-SNAPSHOT/phoenicis-repository-5.0-SNAPSHOT.pom
[WARNING] The POM for org.phoenicis:phoenicis-repository:jar:5.0-SNAPSHOT is missing, no dependency information available
Downloading from jitpack.io: https://jitpack.io/org/phoenicis/phoenicis-scripts/5.0-SNAPSHOT/maven-metadata.xml
Downloading from jitpack.io: https://jitpack.io/org/phoenicis/phoenicis-scripts/5.0-SNAPSHOT/phoenicis-scripts-5.0-SNAPSHOT.pom
[WARNING] The POM for org.phoenicis:phoenicis-scripts:jar:5.0-SNAPSHOT is missing, no dependency information available
Downloading from jitpack.io: https://jitpack.io/org/phoenicis/phoenicis-library/5.0-SNAPSHOT/maven-metadata.xml
Downloading from jitpack.io: https://jitpack.io/org/phoenicis/phoenicis-library/5.0-SNAPSHOT/phoenicis-library-5.0-SNAPSHOT.pom
[WARNING] The POM for org.phoenicis:phoenicis-library:jar:5.0-SNAPSHOT is missing, no dependency information available
Downloading from jitpack.io: https://jitpack.io/org/phoenicis/phoenicis-engines/5.0-SNAPSHOT/maven-metadata.xml
Downloading from jitpack.io: https://jitpack.io/org/phoenicis/phoenicis-engines/5.0-SNAPSHOT/phoenicis-engines-5.0-SNAPSHOT.pom
[WARNING] The POM for org.phoenicis:phoenicis-engines:jar:5.0-SNAPSHOT is missing, no dependency information available
Downloading from jitpack.io: https://jitpack.io/org/phoenicis/phoenicis-containers/5.0-SNAPSHOT/maven-metadata.xml
Downloading from jitpack.io: https://jitpack.io/org/phoenicis/phoenicis-containers/5.0-SNAPSHOT/phoenicis-containers-5.0-SNAPSHOT.pom
[WARNING] The POM for org.phoenicis:phoenicis-containers:jar:5.0-SNAPSHOT is missing, no dependency information available
Downloading from jitpack.io: https://jitpack.io/org/phoenicis/phoenicis-settings/5.0-SNAPSHOT/maven-metadata.xml
Downloading from jitpack.io: https://jitpack.io/org/phoenicis/phoenicis-settings/5.0-SNAPSHOT/phoenicis-settings-5.0-SNAPSHOT.pom
[WARNING] The POM for org.phoenicis:phoenicis-settings:jar:5.0-SNAPSHOT is missing, no dependency information available
Downloading from jitpack.io: https://jitpack.io/org/phoenicis/phoenicis-configuration/5.0-SNAPSHOT/phoenicis-configuration-5.0-SNAPSHOT.jar
Downloading from jitpack.io: https://jitpack.io/org/phoenicis/phoenicis-library/5.0-SNAPSHOT/phoenicis-library-5.0-SNAPSHOT.jar
Downloading from jitpack.io: https://jitpack.io/org/phoenicis/phoenicis-scripts/5.0-SNAPSHOT/phoenicis-scripts-5.0-SNAPSHOT.jar
Downloading from jitpack.io: https://jitpack.io/org/phoenicis/phoenicis-repository/5.0-SNAPSHOT/phoenicis-repository-5.0-SNAPSHOT.jar
Downloading from jitpack.io: https://jitpack.io/org/phoenicis/phoenicis-engines/5.0-SNAPSHOT/phoenicis-engines-5.0-SNAPSHOT.jar
Downloading from jitpack.io: https://jitpack.io/org/phoenicis/phoenicis-containers/5.0-SNAPSHOT/phoenicis-containers-5.0-SNAPSHOT.jar
Downloading from jitpack.io: https://jitpack.io/org/phoenicis/phoenicis-settings/5.0-SNAPSHOT/phoenicis-settings-5.0-SNAPSHOT.jar
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  8.687 s
[INFO] Finished at: 2019-06-11T21:18:45+02:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal on project phoenicis-javafx: Could not resolve dependencies for project org.phoenicis:phoenicis-javafx:jar:5.0-SNAPSHOT: The following artifacts could not be resolved: org.phoenicis:phoenicis-configuration:jar:5.0-SNAPSHOT, org.phoenicis:phoenicis-repository:jar:5.0-SNAPSHOT, org.phoenicis:phoenicis-scripts:jar:5.0-SNAPSHOT, org.phoenicis:phoenicis-library:jar:5.0-SNAPSHOT, org.phoenicis:phoenicis-engines:jar:5.0-SNAPSHOT, org.phoenicis:phoenicis-containers:jar:5.0-SNAPSHOT, org.phoenicis:phoenicis-settings:jar:5.0-SNAPSHOT: Could not find artifact org.phoenicis:phoenicis-configuration:jar:5.0-SNAPSHOT in jitpack.io (https://jitpack.io) -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/DependencyResolutionException
```

Comment (qparis) [reply phoenicis-scripts/src/main/java/org/phoenicis/scripts/wizard/UiSetupWizardImplementation.java]: Maven is trying to download phoenicis. I suspect that it is not properly installed inside .m2

**Pull request merged**